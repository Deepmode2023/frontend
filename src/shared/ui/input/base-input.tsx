import TextField, { StandardTextFieldProps } from "@mui/material/TextField";
import { UseFormRegisterReturn, FieldError } from "react-hook-form";
import { motion, AnimatePresence } from "framer-motion";

export interface IBaseInput<TNameField extends string>
  extends StandardTextFieldProps {
  registerOptions: UseFormRegisterReturn<TNameField>;
  isRequired?: boolean;
  label: string;
  formError: FieldError | undefined;
}

const variants = {
  initial: {
    y: "-15px",
    opacity: 0,
    backgroundColor: "red",
  },
  active: {
    opacity: 1,
    y: 0,
    backgroundColor: "white",
  },
  inactive: {
    opacity: 0,
    y: "-5px",
  },
};

function BaseInput<TNameField extends string>({
  registerOptions,
  label,
  formError,
  isRequired = false,
  ...inputProps
}: IBaseInput<TNameField>) {
  return (
    <div data-testid="base_input" className="flex flex-col gap-1">
      <TextField
        id="input-with-sx"
        label={label}
        variant="standard"
        {...registerOptions}
        {...inputProps}
      />
      <AnimatePresence>
        {formError?.message && (
          <motion.span
            variants={variants}
            initial={"initial"}
            animate={"active"}
            exit={"inactive"}
            className="text-red-600"
            data-testid="base_input_message"
            transition={{
              type: "spring",
              damping: 10,
              stiffness: 100,
              duration: 0.5,
            }}
          >
            {formError.message}
          </motion.span>
        )}
      </AnimatePresence>
    </div>
  );
}

export { BaseInput };
