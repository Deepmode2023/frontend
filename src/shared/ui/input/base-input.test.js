import { describe, test, expect, jest, beforeEach } from "bun:test";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BaseInput } from "./base-input";

import * as matchers from "@testing-library/jest-dom/matchers";

expect.extend(matchers);

describe("BaseInput component", () => {
  let component;
  const registerOptions = {
    name: "testName",
    onChange: jest.fn(),
    onBlur: jest.fn(),
    ref: jest.fn(),
  };

  beforeEach(() => {
    component = render(
      <BaseInput
        label="Test Label"
        registerOptions={registerOptions}
        formError={undefined}
      />
    );
  });

  test("... & check with default display", () => {
    expect(screen.getByLabelText("Test Label")).toBeInTheDocument();
  });

  test("... & passes the props correctly to TextField", () => {
    const { container } = component;
    const input = container.querySelector("input");

    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute("name", "testName");
  });

  test("... & displays error message when formError is present", async () => {
    const formError = { message: "This is an error message" };
    const { rerender } = component;
    await waitFor(() => {
      rerender(
        <BaseInput
          label="Test Label"
          registerOptions={registerOptions}
          formError={formError}
        />
      );
    });
    const messageComponent = screen.getByTestId("base_input_message");
    expect(messageComponent).toBeInTheDocument();
    expect(messageComponent).toHaveTextContent(formError.message);

    expect(component.container.innerHTML).toMatchSnapshot();
  });

  test("... & handles user input correctly", async () => {
    const input = component.container.querySelector("input");
    await userEvent.type(input, "Hello");

    expect(registerOptions.onChange).toHaveBeenCalledTimes(5);
  });

  test("... & make snapshot", () => {
    expect(component.container.innerHTML).toMatchSnapshot();
  });
});
