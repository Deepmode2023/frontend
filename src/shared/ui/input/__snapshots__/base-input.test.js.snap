// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseInput component ... & make snapshot 1`] = `"<div data-testid="base_input" class="flex flex-col gap-1"><div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-19fz36k-MuiFormLabel-root-MuiInputLabel-root" data-shrink="false" for="input-with-sx" id="input-with-sx-label">Test Label</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-r5in8e-MuiInputBase-root-MuiInput-root"><input aria-invalid="false" id="input-with-sx" name="testName" type="text" class="MuiInputBase-input MuiInput-input css-1o3an4k-MuiInputBase-input-MuiInput-input" value=""></div></div></div>"`;

exports[`BaseInput component ... & displays error message when formError is present 1`] = `"<div data-testid="base_input" class="flex flex-col gap-1"><div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"><label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-19fz36k-MuiFormLabel-root-MuiInputLabel-root" data-shrink="false" for="input-with-sx" id="input-with-sx-label">Test Label</label><div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-r5in8e-MuiInputBase-root-MuiInput-root"><input aria-invalid="false" id="input-with-sx" name="testName" type="text" class="MuiInputBase-input MuiInput-input css-1o3an4k-MuiInputBase-input-MuiInput-input" value=""></div></div><span class="text-red-600" data-testid="base_input_message" style="opacity: 0; background-color: white; transform: translateY(-15px) translateZ(0);">This is an error message</span></div>"`;
